File: alsh_mips/src/cpp/lsh.cpp
#include "lsh.h"

#include <random>
#include <iostream>
#include <math.h>
#include <sstream>
#include <cmath>

using namespace std;

std::string hash_combine(const std::vector<int> &hashes) {
  int size = hashes.size();
  std::stringstream ss;

  for (int i = 0; i < size; ++i) {
    ss << hashes.at(i) << ",";
  }

  std::string str;
  ss >> str;
  return  str;
}

std::string cosine_hash_combine(const std::vector<int> &hashes) {
  // return sum(2**i if h > 0 else 0 for i, h in enumerate(hashes))

  long s = 0;
  for (int i = 0; i < hashes.size(); ++i) {
    if(hashes[i]>0)
      s += pow(2,i);
  }

  string str = to_string(s);
  return  str;
}


double dot(const VectorXd &u, const VectorXd &v) {
  return u.adjoint()*v;
}

VectorXd g_ext_norm(const VectorXd &vec, const int m) {
  double l2norm_square = dot(vec, vec);
  VectorXd res(m);
  for(int i=0;i<m;i++) {
    res(i) = pow(l2norm_square,i+1);
  }

  return res;
}

VectorXd g_ext_half(const int m) {
  VectorXd res(m);
  for(int i=0;i<m;i++)  {
    res(i) = 0.5;
  }

  return res;
}

VectorXd g_ext_norm_cosine(const VectorXd &vec, const int m) {
  double l2norm_square = dot(vec, vec);
  VectorXd res(m);

  for(int i=0;i<m;i++) {
    res(i) = 0.5 - pow(l2norm_square, i+1);
  }

  return res;
}

VectorXd g_ext_norm_simple(const VectorXd &vec, const int m) {
  double l2norm_square = dot(vec, vec);
  VectorXd res(m);

  for(int i=0;i<m;i++) {
    res(i) = sqrt(1 - l2norm_square);
  }

  return res;
}

VectorXd g_ext_zero(const int m) {
  VectorXd res(m);
  for(int i=0;i<m;i++) {
    res(i) = 0;
  }

  return res;
}

MatrixXd g_index_extend(const MatrixXd &datas, const int m) {   //TODO why 2*m
  MatrixXd res(datas.rows(),datas.cols()+2*m);

  for(int i=0;i<datas.rows();i++) {
    res.row(i).head(datas.cols()) = datas.row(i);
    res.row(i).segment(datas.cols(),m) = g_ext_norm(datas.row(i), m);
    res.row(i).tail(m) = g_ext_half(m);
  }

  return res;
}

MatrixXd g_query_extend(const MatrixXd &queries, const int m) {
  MatrixXd res(queries.rows(),queries.cols()+2*m);

  for(int i=0;i<queries.rows();i++) {
    res.row(i).head(queries.cols()) = queries.row(i);
    res.row(i).segment(queries.cols(),m) = g_ext_half(m);
    res.row(i).tail(m) = g_ext_norm(queries.row(i), m);
  }

  return res;
}

MatrixXd g_index_cosine_extend(const MatrixXd &datas, const int m) {
  MatrixXd res(datas.rows(),datas.cols()+2*m);

  for(int i=0;i<datas.rows();i++) {
    res.row(i).head(datas.cols()) = datas.row(i);
    res.row(i).segment(datas.cols(),m) = g_ext_norm_cosine(datas.row(i), m);
    res.row(i).tail(m) = g_ext_zero(m);
  }

  return res;
}

MatrixXd g_query_cosine_extend(const MatrixXd &queries, const int m) {
  MatrixXd res(queries.rows(),queries.cols()+2*m);

  for(int i=0;i<queries.rows();i++) {
    res.row(i).head(queries.cols()) = queries.row(i);
    res.row(i).segment(queries.cols(),m) = g_ext_zero(m);
    res.row(i).tail(m) = g_ext_norm_cosine(queries.row(i), m);
  }

  return res;
}

MatrixXd g_index_simple_extend(const MatrixXd &datas, const int m) {
  assert (1 == m);
  MatrixXd res(datas.rows(),datas.cols()+2);

  for(int i=0;i<datas.rows();i++) {
    res.row(i).head(datas.cols()) = datas.row(i);
    res.row(i).segment(datas.cols(),m) = g_ext_norm_simple(datas.row(i), m);
    res.row(i).tail(m) = g_ext_zero(m);
  }

  return res;
}

MatrixXd g_query_simple_extend(const MatrixXd &queries, const int m) {
  assert (1 == m);
  MatrixXd res(queries.rows(),queries.cols()+2);

  for(int i=0;i<queries.rows();i++) {
    res.row(i).head(queries.cols()) = queries.row(i);
    res.row(i).segment(queries.cols(),m) = g_ext_zero(m);
    res.row(i).tail(m) = g_ext_norm_simple(queries.row(i), m);
  }

  return res;
}

double g_max_norm(const MatrixXd &datas) {
  VectorXd data_row(datas.rows());

  for(int i=0;i<datas.rows();i++) {
    data_row(i) = datas.row(i).norm();
  }

  return data_row.maxCoeff();
}

// datas transformation. S(xi) = (U / M) * xi
MatrixXd g_transformation(const MatrixXd &datas, double &ratio, double &max_norm) {
  // U < 1  ||xi||2 <= U <= 1. recommend for 0.83
  const double U = 0.83;

  //U = 0.75
  max_norm = g_max_norm(datas);
  ratio = double(U / max_norm);
  MatrixXd res(datas.rows(), datas.cols());

  for(int i=0;i<datas.rows();i++) {
    res.row(i) = datas.row(i) * ratio;
  }

  return res;
}

// normalization for each query
MatrixXd g_normalization(const MatrixXd &queries) {
  const double U = 0.83;

  //U = 0.75
  MatrixXd norm_queries(queries.rows(), queries.cols());
  double ratio;

  for(int i=0;i<queries.rows();i++) {
    ratio = double(U / queries.row(i).norm());
    norm_queries.row(i) = queries.row(i) * ratio;
  }

  return norm_queries;
}

double L2Lsh_distance(const VectorXd &u, const VectorXd &v) {
  return sqrt((u - v).squaredNorm());
}

double ConsineLsh_distance(const VectorXd &u, const VectorXd &v) {
  return 1 - (dot(u, v) / sqrt(dot(u, u) * dot(v, v)));
}

L2Lsh::L2Lsh(double r, int d) {
  srand((unsigned int) time(0));
  r_ = r;
  d_ = d;
  VectorXd v = VectorXd::Random(1);
  b_ = v(0) * r_ / 2 ;        // 0 < b < r    TODO

  std::random_device rd;
  std::mt19937 e2(rd());

  //gauss gen
  std::normal_distribution<> distribution(0, 1.0);
  auto gauss = [&] () {return distribution(e2);};
  data_ = VectorXd::NullaryExpr(d, gauss);
}

L2Lsh::~L2Lsh() {
}

int L2Lsh::init(double r, int d) {
  return 0;
}

int L2Lsh::hash(const VectorXd &vec) {
  return (int)((dot(vec, data_) + b_) / r_);
}

double L2Lsh::distance(const VectorXd &u, const VectorXd &v) {
  return sqrt((u - v).squaredNorm());
}

ConsineLsh::ConsineLsh(double r, int d) {
  srand((unsigned int) time(0));
  d_ = d;

  // gauss gen
  std::random_device rd;
  std::mt19937 e2(rd());
  std::normal_distribution<> distribution(0, 1);
  auto gauss = [&] () {return distribution(e2);};
  data_ = VectorXd::NullaryExpr(d, gauss);
}

ConsineLsh::~ConsineLsh() {
}

int ConsineLsh::init(double r, int d) {
  return 0;
}

int ConsineLsh::hash(const VectorXd &vec) {
  double dot_val = dot(vec, data_);

  if (dot_val > 0) {
      return 1;
  }

  return 0;
}

double ConsineLsh::distance(const VectorXd &u, const VectorXd &v) {
  return 1 - (dot(u, v) / sqrt(dot(u, u) * dot(v, v)));
}

File: alsh_mips/src/python/lsh.py
# -*- coding: utf-8 -*-

"""
# Locality Sensitive Hashing for MIPS

## Reference:

- LSH:
[Locality-Sensitive Hashing Scheme Based on p-Stable Distributions](http://www.cs.princeton.edu/courses/archive/spr05/cos598E/bib/p253-datar.pdf)

- L2-ALSH:
[Asymmetric LSH (ALSH) for Sublinear Time Maximum Inner Product Search (MIPS)](https://papers.nips.cc/paper/5329-asymmetric-lsh-alsh-for-sublinear-time-maximum-inner-product-search-mips.pdf)

- Sign-ALSH(Cosine):
[Improved Asymmetric Locality Sensitive Hashing (ALSH) for Maximum Inner Product Search (MIPS)](http://auai.org/uai2015/proceedings/papers/96.pdf)

- Simple-ALSH(Cosine):
[On Symmetric and Asymmetric LSHs for Inner Product Search](https://arxiv.org/abs/1410.5518)

"""

import math
import random
from abc import ABCMeta, abstractmethod
from collections import defaultdict
from operator import itemgetter
from global_mgr import gol

import numpy as np

# matrix inner product
def dot(u, v):
    # return sum(ux * vx for ux, vx in zip(u,v))
    return np.dot(u, v)

def g_ext_norm(vec, m):
    l2norm_square = dot(vec, vec)
    return [l2norm_square**(i+1) for i in xrange(m)]

def g_ext_half(m):
    return [0.5 for i in xrange(m)]

def g_ext_norm_cosine(vec, m):
    l2norm_square = dot(vec, vec)
    return [0.5 - l2norm_square**(i+1) for i in xrange(m)]

def g_ext_norm_simple(vec, m):
    l2norm_square = dot(vec, vec)
    return [np.sqrt(1 - l2norm_square) for i in xrange(m)]

def g_ext_zero(m):
    return [0 for i in xrange(m)]

# [x] => [x;    ||x||**2; ||x||**4; ...; ||x||**(2*m);    1/2; ...; 1/2(m)]
def g_index_extend(datas, m):
    return [(dv + g_ext_norm(dv, m) + g_ext_half(m)) for dv in datas]

# [x] => [x;    1/2; ...; 1/2(m);    ||x||**2; ||x||**4; ...; ||x||**(2*m)]
def g_query_extend(queries, m):
    return [(qv + g_ext_half(m) + g_ext_norm(qv, m)) for qv in queries]

# [x] => [x;    1/2 - ||x||**2; 1/2 - ||x||**4; ...; 1/2 - ||x||**(2*m);    0; ...; 0(m)]
def g_index_cosine_extend(datas, m):
    return [(dv + g_ext_norm_cosine(dv, m) + g_ext_zero(m)) for dv in datas]

# [x] => [x;    0; ...; 0(m);    1/2 - ||x||**2; 1/2 - ||x||**4; ...; 1/2 - ||x||**(2*m)]
def g_query_cosine_extend(queries, m):
    return [(qv + g_ext_zero(m) + g_ext_norm_cosine(qv, m)) for qv in queries]

# [x] => [x;    sqrt(1 - ||x||**2);    0]
def g_index_simple_extend(datas, m):
    assert (1 == m)
    return [(dv + g_ext_norm_simple(dv, m) + g_ext_zero(m)) for dv in datas]

# [x] => [x;    0;    sqrt(1 - ||x||**2)]
def g_query_simple_extend(queries, m):
    assert (1 == m)
    return [(qv + g_ext_zero(m) + g_ext_norm_simple(qv, m)) for qv in queries]

# get max norm for two-dimension list
def g_max_norm(datas):
    norm_list = [math.sqrt(dot(dd, dd)) for dd in datas]
    return max(norm_list)

# datas transformation. S(xi) = (U / M) * xi
def g_transformation(datas):
    # U < 1  ||xi||2 <= U <= 1. recommend for 0.83
    U = 0.83
    #U = 0.75
    max_norm = g_max_norm(datas)
    ratio = float(U / max_norm)
    return ratio, max_norm, [[ratio * dx for dx in dd] for dd in datas]

# normalization for each query
def g_normalization(queries):
    U = 0.83
    #U = 0.75
    norm_queries = []
    for qv in queries:
        norm = math.sqrt(dot(qv, qv))
        ratio = float(U / norm)
        norm_queries.append([ratio * qx for qx in qv])
    return norm_queries

class Hash:
    'hash base'

    __metaclass__ = ABCMeta

    def __init__(self):
        pass

    @abstractmethod
    def hash(self, vec):
        print "Hash.hash()"
        pass

    @staticmethod
    def distance(u, v):
        print "Hash.distance()"
        pass

    @staticmethod
    def combine(hashes):
        # print "Hash.combine()"
        return str(hashes)

class L2Lsh(Hash):
    'L2 LSH'

    # r: fixed size
    # d: data length
    # RandomData: random data vector
    def __init__(self, r, d):
        self.r = r
        self.d = d
        self.b = random.uniform(0, self.r)      # 0 < b < r
        self.Data = [random.gauss(0, 1) for i in xrange(self.d)]

        if True == gol.get_value('DEBUG'):
            print 'L2Lsh:\tr: ' + str(self.r) + '\td: ' + str(self.d) + '\tb: ' + str(self.b)
            print 'L2Lsh Data: '
            print self.Data

    def hash(self, vec):
        # use str() as a naive way of forming a single value
        return int((dot(vec, self.Data) + self.b) / self.r)

    # Euclidean Distance
    @staticmethod
    def distance(u, v):
        # print "L2Lsh.distance()"
        """
        u_na = np.array(u)
        v_na = np.array(v)
        return np.linalg.norm(u_na - v_na)
        """
        return sum((ux - vx)**2 for ux, vx in zip(u, v))**0.5

class CosineLsh(Hash):
    'Cosine LSH'

    def __init__(self, d):
        self.d = d
        self.Data = [random.gauss(0, 1) for i in xrange(self.d)]

        if True == gol.get_value('DEBUG'):
            print 'CosineLsh:\td: ' + str(self.d)
            print 'CosineLsh Data: '
            print self.Data

    def hash(self, vec):
        return int(0 < dot(vec, self.Data))

    # Cosine Distance
    @staticmethod
    def distance(u, v):
        # print "Cosine.distance()"
        return 1 - (dot(u, v) / (dot(u, u) * dot(v, v))**0.5)

    @staticmethod
    def combine(hashes):
        # print "Cosine.combine()"
        return sum(2**i if h > 0 else 0 for i, h in enumerate(hashes))
